packages/protocol/contracts/bridge/Bridge.sol

```diff
@@ -238,106 +262,106 @@ contract Bridge is EssentialContract, IBridge {
         sameChain(_message.destChainId)
         nonReentrant
     {
-        bytes32 msgHash = hashMessage(_message);
-        if (messageStatus[msgHash] != Status.NEW) revert B_STATUS_MISMATCH();
+        Local memory local;
+        local.gas = gasleft();
+        local.notProcessedByOwner = msg.sender != _message.destOwner;
+
+        // If the gas limit is set to zero, only the owner can process the message.
+        if (local.notProcessedByOwner) {
+            if (_message.gasLimit == 0) revert B_PERMISSION_DENIED();
+            if (local.gas < _message.gasLimit) revert B_INVALID_GAS_LIMIT();
+        }
+
+        ProofReceipt memory receipt;
+        (local.msgHash, receipt) = _checkStatusAndReceipt(_message);
+        local.signalService = resolve(LibStrings.B_SIGNAL_SERVICE, false);
+        local.invocationDelay = getInvocationDelay();
 
-        address signalService = resolve(LibStrings.B_SIGNAL_SERVICE, false);
+        if (receipt.receivedAt == 0) {
+            bool received = _proveSignalReceived(
+                local.signalService, local.msgHash, _message.srcChainId, _proof
+            );
+            if (!received) revert B_NOT_RECEIVED();
 
-        uint64 receivedAt = proofReceipt[msgHash].receivedAt;
-        if (receivedAt == type(uint64).max) revert B_MESSAGE_SUSPENDED();
+            receipt = ProofReceipt(uint64(block.timestamp), 0, 0);
 
-        (uint256 invocationDelay, uint256 invocationExtraDelay) = getInvocationDelays();
+            if (local.invocationDelay != 0) {
+                if (local.notProcessedByOwner) {
+                    receipt.gasUsed =
+                        uint32(local.gas - gasleft() + GAS_OVERHEAD_RECEIVING + _proof.length >> 4);
```

Comment:
`abi.decode` allows trailing bytes, so this attack would currently be possible. I'm not sure how this problem can be fixed. In smart wallets I think this is mostly a non-issue because most of the data is either fixed or needs to be signed by the wallet owner, in this case however the proof is just sent by the relayer, so it's tricky.

Reply:
It's actually even more complex than I thought because of the caching. The proof verification cost is now paid by the user, which makes sense, but the caching which is decided by the relayer is also paid by the user now. So there isn't much reason for the relayer to use the caching unless that somehow increases its profit (which could still be the case if it allows extracting a higher fee from the user).

So that also removed the benefit of the relayer being able to lower the cost of all bridge transactions by batching multiple togethe, lowering the cost of all bridge transactions in the batch. Now this caching cost (if even done) would be paid by the bridge transaction doing the caching (instead of, in the best case, the cost being charged spread out over the multiple message).

I think the main cause of these problems is the shared gas limit. There are two very distinct things the user wants the relayer to do, which are the two different steps. I think we need to have a gas limit/max fee per step to be able to handle the two steps correctly (even when not done in two separate steps):

- Step 1 - verifying the merkle proof + proof data: fixed fee not taking the gas used in account (still based on the base fee, so lower base fee results in less of the max fee being used, the fee paid is `message.maxFeeProofVerification * (block.basefee / message.maxBaseFee)`. This ensures that the relayer will want to optimize the proof verification as much as possible (making a profit when doing so), and also makes sure that the user can pay a fee that forces the relayer to optimize the processing of its bridge transactions with caching by setting the max fee low.
- Step 2 - message call invocation: basically the system in this PR based on gas used.

I think something like this would work, but needs some more thinking and eyes on it.